internal/scheduler/ingester.go

```diff
@@ -0,0 +1,777 @@
+
+		// Write batches to postgres and, for each batch, ack all pulsar messages the batch was made up of.
+		g.Go(func() error { return srv.ProcessBatches(ctx, batchChannel) })
+
+		// Run pipeline until any error.
+		err = g.Wait()
+		// TODO: Detect recoverable errors.
+		if err != nil {
+			// Unrecoverable error; exit.
+			return err
+		}
+	}
+}
+
+func (srv *Ingester) ProcessMessages(ctx context.Context, messageChannel <-chan pulsar.Message, batchChannel chan<- *Batch) error {
+	log := ctxlogrus.Extract(ctx)
+
+	// In-progress batches.
+	batches := make([]*Batch, 0)
+
+	// Ticker to trigger periodically writing to postgres.
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	// Run until ctx is canceled.
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case msg := <-messageChannel:
+
+			// Incoming gRPC requests are annotated with a unique id,
+			// which is included with the corresponding Pulsar message.
+			requestId := pulsarrequestid.FromMessageOrMissing(msg)
+
+			// Put the requestId into a message-specific context and logger,
+			// which are passed on to sub-functions.
+			messageCtx, ok := requestid.AddToIncomingContext(ctx, requestId)
+			if !ok {
+				messageCtx = ctx
+			}
+			messageLogger := log.WithFields(logrus.Fields{"messageId": msg.ID(), requestid.MetadataKey: requestId})
+			ctxWithLogger := ctxlogrus.ToContext(messageCtx, messageLogger)
+
+			// Any consumer can seek on any topic partition.
+			// An unexpected seek may cause messages to be delivered out of order,
+			// i.e., a received messages may not be newer than the previous message.
+			//
+			// We compare the id of each received message with that of the previous message
+			// on the same partition to detect such out-of-order messages.
+			//
+			// Note that, while possible, this should be a rare occurrence.
+			//
+			// TODO: We should check for this and restart the consumer if it happens.
+			partitionIdx := msg.ID().PartitionIdx()
+			if srv.lastMessageIdByPartition == nil {
+				srv.lastMessageIdByPartition = make(map[int32]pulsar.MessageID)
+			}
+			if lastMessageId, ok := srv.lastMessageIdByPartition[partitionIdx]; ok {
+				msgIsOufOfOrder, err := pulsarutils.FromMessageId(lastMessageId).GreaterEqual(msg.ID())
+				if err != nil {
+					return err
+				}
+				if msgIsOufOfOrder {
+					// pulsarutils.PulsarMessageId prints nicely, so we convert to those.
+					return fmt.Errorf(
+						"unexpected seek detected: received messages out of order for topic %s: id of received message is %s, but the previous message has id %s",
+						msg.Topic(), pulsarutils.FromMessageId(msg.ID()), pulsarutils.FromMessageId(lastMessageId),
+					)
+				}
+			}
+			srv.lastMessageIdByPartition[partitionIdx] = msg.ID()
+
+			var err error
+			batches, err = srv.ProcessMessage(ctxWithLogger, msg, batches)
+			if err != nil {
+				logging.WithStacktrace(messageLogger, err).Warnf("processing message failed; ignoring")
```

Comment:
Do you want a `continue` here? I guess it'll ptrobably work because `batches` will probably be nil if there' s an errr here, but I think it'd be safer not to use a result from a function that's returned an error.